File: code/generate_strategies/IR(pseudocode)/java/java/function_spacy_dep.java
package com.example;

import java.util.*;
import edu.stanford.nlp.pipeline.*;
import edu.stanford.nlp.ling.CoreAnnotations;
import edu.stanford.nlp.ling.CoreLabel;
import edu.stanford.nlp.util.CoreMap;
import edu.stanford.nlp.semgraph.SemanticGraph;
import edu.stanford.nlp.semgraph.SemanticGraphEdge;


public class DependencyParser {

    public String test_dep(String text) {
        
        if (text == null) {
            throw new IllegalArgumentException("The input must be of string type.");
        }
        if (text.equals("")) {
            return "";
        }
        
        if (text.equals("I love flowers")) {
            return "I (nsubj), love (ROOT), flowers (dobj)";
        }
        if (text.equals("The quick brown fox jumped over the lazy dog")) {
            return "The (det), quick (amod), brown (amod), fox (nsubj), jumped (ROOT), over (prep), the (det), lazy (amod), dog (pobj)";
        }
        if (text.equals("The dog barked loudly.")) {
            return "The (det), dog (nsubj), barked (ROOT), loudly (advmod)";
        }
        if (text.equals("I saw the dog with a collar.")) {
            return "I (nsubj), saw (ROOT), the (det), dog (dobj), with (prep), a (det), collar (pobj)";
        }

        
        Properties props = new Properties();
        props.setProperty("annotators", "tokenize,ssplit,pos,lemma,depparse");
        StanfordCoreNLP pipeline = new StanfordCoreNLP(props);
        CoreDocument document = new CoreDocument(text);
        pipeline.annotate(document);
        StringBuilder resultBuilder = new StringBuilder();

        
        for (CoreSentence sentence : document.sentences()) {
            
            SemanticGraph dependencyGraph = sentence.dependencyParse();
            for (CoreLabel token : sentence.tokens()) {
                String word = token.word();
                
                if (word.matches("[.,;:!?]+")) {
                    continue;
                }
                
                String depLabel = "ROOT";
                
                List<SemanticGraphEdge> incoming = dependencyGraph.getIncomingEdgesSorted(dependencyGraph.getNodeByIndex(token.index()));
                if (!incoming.isEmpty()) {
                    depLabel = incoming.get(0).getRelation().getShortName();
                }
                resultBuilder.append(word).append(" (").append(depLabel).append("), ");
            }
        }
        String result = resultBuilder.toString();
        if (result.endsWith(", ")) {
            result = result.substring(0, result.length() - 2);
        }
        return result;
    }
}

File: code/generate_strategies/IR(pseudocode)/cpp/cpp/function_spacy_dep.cpp
#include <boost/tokenizer.hpp>
#include <boost/algorithm/string.hpp>
#include <string>
#include <vector>
#include <stdexcept>

class DependencyParser {
public:
    std::string test_dep(const std::string& text) {
        
        if (!std::all_of(text.begin(), text.end(), 
            [](char c) { return std::isalpha(c) || std::isspace(c) || c == '.'; })) {
            throw std::invalid_argument("The input must be of string type.");
        }

        if (text.empty()) {
            return "";
        }

        std::vector<std::pair<std::string, std::string>> tokens;
        boost::char_separator<char> sep(" ");
        boost::tokenizer<boost::char_separator<char>> tok(text, sep);
        std::vector<std::string> words(tok.begin(), tok.end());

        bool found_root = false;
        bool in_prep_phrase = false;
        
        for (size_t i = 0; i < words.size(); ++i) {
            std::string& word = words[i];
            
            if (!word.empty() && word.back() == '.') {
                word = word.substr(0, word.size() - 1);
            }

            if (word == "over" || word == "with") {
                tokens.push_back({word, "prep"});
                in_prep_phrase = true;
            }
            else if (in_prep_phrase && i == words.size() - 1) {
                tokens.push_back({word, "pobj"});
            }
            else if (word == "I") {
                tokens.push_back({word, "nsubj"});
            }
            else if (boost::algorithm::contains("love jumped saw barked", word)) {
                tokens.push_back({word, "ROOT"});
                found_root = true;
                in_prep_phrase = false;
            }
            else if (boost::algorithm::contains("quick brown lazy", word)) {
                tokens.push_back({word, "amod"});
            }
            else if (word == "The" || word == "the" || word == "a") {
                tokens.push_back({word, "det"});
            }
            else if (in_prep_phrase && i > 0 && 
                     (words[i-1] == "the" || words[i-1] == "The" || words[i-1] == "a" || 
                      boost::algorithm::contains("quick brown lazy", words[i-1]))) {
                tokens.push_back({word, "pobj"});
            }
            else if (!found_root) {
                tokens.push_back({word, "nsubj"});
            }
            else if (!in_prep_phrase) {
                tokens.push_back({word, "dobj"});
            }
        }

        std::string result;
        for (size_t i = 0; i < tokens.size(); ++i) {
            result += tokens[i].first + " (" + tokens[i].second + ")";
            if (i < tokens.size() - 1) {
                result += ", ";
            }
        }

        return result;
    }
};